Feature: CPI Events API

ngundotra opened the pull request:
Adds `emit_cpi` to allow programs to store logs in CPI data, the same way Metaplex's compressed NFTs do.

Closes #2408

ngundotra committed:
add permissionless event cpi api

ngundotra committed:
add cpi event test

armaniferrante commented on lang/src/lib.rs:
```diff
@@ -193,6 +194,26 @@ pub trait Event: AnchorSerialize + AnchorDeserialize + Discriminator {
+    }};
+}
+
+pub fn _emit_cpi_invoke(ix_data: Vec<u8>, program: &AccountInfo) -> Result<()> {
```
can you put this inside the __private module below?

ngundotra replied:
@armaniferrante fixed

ngundotra committed:
move __emit_cpi_invoke to __private in lib.rs

armaniferrante commented on lang/syn/src/codegen/event.rs:
```diff
@@ -0,0 +1 @@
+
```
empty file, please remove

ngundotra replied:
Removed! Thanks

ngundotra committed:
export emit_cpi and _emit_cpi_data in prelude

ngundotra committed:
remove empty file

ngundotra commented:
Unfortunately I cannot move the `_emit_cpi_data` and `emit_cpi` macro into the `lang/attribute` crate because `attribute` crate is a `proc_macro` crate, and `emit_cpi` is a declarative macro, which are incompatible exports. ie `attribute` can only export proc_macro macros 🙃

ngundotra committed:
rewrite emit_cpi as a proc_macro

ngundotra commented:
Rewrote `emit_cpi` as a `proc_macro` that takes a tuple `(AccountInfo, <dyn anchor_lang::Event>)`

ngundotra committed:
remove unused code

armaniferrante commented on lang/src/lib.rs:
```diff
@@ -358,6 +361,17 @@ pub mod __private {
     }
+
+    #[doc(hidden)]
+    pub fn _emit_cpi_invoke(ix_data: Vec<u8>, program: &AccountInfo) -> anchor_lang::Result<()> {
```
Don't mean to nit, but what's the advantage of having this extra function vs just inling this code in the one place it's used above?

ngundotra replied:
I thought it was cleaner development, really no advantage/disadvantage that I know of off top

armaniferrante replied:
Let's inline it, instead. The reasoning is that removes this __private mod hack and avoid a file switch for first time readers of the code.

ngundotra committed:
inline _emit_cpi_invoke to proc_macro declaration

armaniferrante commented on lang/syn/src/codegen/program/handlers.rs:
```diff
@@ -91,6 +91,16 @@ pub fn generate(program: &Program) -> proc_macro2::TokenStream {
+            #[inline(never)]
+            #[cfg(not(feature = "no-cpi-events"))]
+            pub fn __event_dispatch(program_id: &Pubkey, accounts: &[AccountInfo], event_data: &[u8]) -> anchor_lang::Result<()> {
+                Ok(())
```
Should we be using the instructions sysvar to check that this call was self-referential, i.e., that the previous instruction was to this program?

Arrowana replied:
An example https://github.com/Ellipsis-Labs/phoenix-v1/blob/ee9cc0333fbab6ee20861e51201951b3e9c38dc9/src/lib.rs#L104-L124

ngundotra replied:
how about two different macros, `emit_cpi!` and `emit_cpi_signed!` ?

I'll look into the introspection stuff. 

I think devs should have options

Arrowana replied:
i would say log authority is better, it is also an account to be provided but much simpler to check.
What would be the advantage of the ix introspection?

ngundotra replied:
I don't understand why `log_authority` is even needed here?

Are people writing indexers that parse `ix`s from TransactionStatus Metadata separately from their transaction context? Is that the norm? I don't understand how `log_authority` helps indexers? Sure it prevents other programs from invoking that instruction, but it will still show up in getSignaturesForAddress. 

Can you provide a concrete example of a program that benefits from having `log_authority` guarding the event ix?

Arrowana replied:
You need to ask @jarry-xiao

jarry-xiao replied:
It’s not 100% necessary, but it prevents footguns (which I think is one of the key tenets of anchor).

We’ve chatted about this before, and I gave a concrete example of making a CPI from a rogue program into the log instruction of the target program. The parser can defend against this but the logic becomes more error prone.

I think it’s fine if you provide an indexing example of how one might process this along with a test case of how it blocks out erroneous instances of calling the log instruction (both through top level transaction calls and CPIs from rogue programs)

jarry-xiao replied:
Also log_authority means you don’t need introspection, which I think is a win. To my knowledge, you still need to explicitly pass in SysvarInstructions

jarry-xiao replied:
Here’s a super concrete example. Suppose you have a target program T and a rogue program R. R first makes a normal call to T, which will emit a regular event. Then in the same instruction it makes a call to T’s log instruction. You can figure out this is erroneous if you have the tx stack depth, but I think it’s ambiguous in the current interface.

In a vacuum you also know nothing about T’s logging patterns. Maybe it calls emit_cpi! a variable number of times. The point is there’s no way to know (again, until stack depth is in the transaction object which could realistically take months)

However if R’s call to T’s log instruction fails 100% of the time, then there’s nothing to ever worry about

Arrowana replied:
Another point for the log authority, if your program allows arbitrary cpi (or only arbitrary self cpi) for any reason (flash loan capability...), then someone can write random logs without the log authority.

acheroncrypto commented on lang/attribute/event/src/lib.rs:
```diff
@@ -81,6 +81,35 @@ pub fn emit(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    proc_macro::TokenStream::from(quote! {
+        let program_info: &anchor_lang::solana_program::account_info::AccountInfo = &#first;
+
+        let __disc: Vec<u8> = crate::event::EVENT_IX_TAG_LE.to_vec();
```
There is no need to convert the array into `Vec` here. Can get rid of this line completely and use `crate::event::EVENT_IX_TAG_LE` instead.

ngundotra replied:
Thanks, fixed

acheroncrypto commented on lang/attribute/event/src/lib.rs:
```diff
@@ -81,6 +81,35 @@ pub fn emit(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+
+        let __disc: Vec<u8> = crate::event::EVENT_IX_TAG_LE.to_vec();
+        let __inner_data: &Vec<u8> = &anchor_lang::Event::data(&#second);
+        let ix_data: Vec<u8> = __disc.iter().chain(__inner_data.iter()).cloned().collect();
```
We can use `into_iter` instead of `iter` to avoid cloning the data.

```rs
let ix_data: Vec<u8> = crate::event::EVENT_IX_TAG_LE
    .into_iter()
    .chain(__inner_data.into_iter())
    .collect();
```

ngundotra replied:
Fixed!

acheroncrypto commented on lang/attribute/event/src/lib.rs:
```diff
@@ -81,6 +81,35 @@ pub fn emit(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+        let program_info: &anchor_lang::solana_program::account_info::AccountInfo = &#first;
+
+        let __disc: Vec<u8> = crate::event::EVENT_IX_TAG_LE.to_vec();
+        let __inner_data: &Vec<u8> = &anchor_lang::Event::data(&#second);
```
Should be able to remove the extra references here:
```rs
let __inner_data: Vec<u8> = anchor_lang::Event::data(#second);
```

ngundotra replied:
Fixed

ngundotra committed:
address acheron feedback

acheroncrypto commented on lang/attribute/event/src/lib.rs:
```diff
@@ -81,6 +81,35 @@ pub fn emit(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
 }
 
+#[proc_macro]
+pub fn emit_cpi(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
```
Can we add documentation for this macro?

ngundotra replied:
Will do once I write this up as an sRFC on forum.solana.com. There's still an additional argument that's necessary for the invoke, which is the Instructions sysvar account.

Once this is ironed out, then yes I will add documentation here

jarry-xiao replied:
Wait if you need SysvarInstructions, then you should just use a log authority PDA

ngundotra replied:
Yeah I agree. Just had to check

acheroncrypto commented on lang/attribute/event/src/lib.rs:
```diff
@@ -81,6 +81,35 @@ pub fn emit(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
 
+#[proc_macro]
+pub fn emit_cpi(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let tuple = parse_macro_input!(input as syn::ExprTuple);
```
It would be better if we took 2 arguments as input instead of a tuple here.

ngundotra replied:
updated! Thanks

acheroncrypto commented on lang/attribute/event/src/lib.rs:
```diff
@@ -81,6 +81,35 @@ pub fn emit(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+        panic!("Expected a tuple with exactly two elements.");
+    }
+
+    let first = &elems[0];
```
Would be great if we could name these variables based on what they are instead of `first` or `second`.

ngundotra replied:
Upated

acheroncrypto commented on lang/attribute/event/src/lib.rs:
```diff
@@ -81,6 +81,35 @@ pub fn emit(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let first = &elems[0];
+    let second = &elems[1];
+
+    proc_macro::TokenStream::from(quote! {
```
The code inside `quote!` will expand in user's instruction and override every clashing variable e.g `program_info` `ix_data` `ix`. Why are some variables prefixed with `__` and some aren't?

ngundotra replied:
Good point, fixed

acheroncrypto commented on lang/attribute/event/src/lib.rs:
```diff
@@ -81,6 +81,35 @@ pub fn emit(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let second = &elems[1];
+
+    proc_macro::TokenStream::from(quote! {
+        let program_info: &anchor_lang::solana_program::account_info::AccountInfo = &#first;
```
If we remove the reference here, we can avoid cloning in `invoke`
```suggestion
        let program_info: anchor_lang::solana_program::account_info::AccountInfo = #first;
```

ngundotra replied:
Thanks, fixed

acheroncrypto commented on lang/attribute/event/src/lib.rs:
```diff
@@ -81,6 +81,35 @@ pub fn emit(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+        let ix_data: Vec<u8> = __disc.into_iter().chain(__inner_data.into_iter()).collect();
+
+        let ix = anchor_lang::solana_program::instruction::Instruction::new_with_bytes(
+            program_info.key.clone(),
```
```suggestion
            *program_info.key,
```

acheroncrypto commented on lang/attribute/event/src/lib.rs:
```diff
@@ -81,6 +81,35 @@ pub fn emit(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+            ix_data.as_ref(),
+            vec![anchor_lang::solana_program::instruction::AccountMeta::new_readonly(*program_info.key, false)],
+        );
+        anchor_lang::solana_program::program::invoke(&ix, &[program_info.clone()])
```
No need to clone here

```suggestion
        anchor_lang::solana_program::program::invoke(&ix, &[program_info])
```

ngundotra replied:
Fixed

acheroncrypto commented on tests/events/package.json:
```diff
@@ -15,5 +15,8 @@
     "test": "anchor test"
+  },
+  "dependencies": {
+    "@coral-xyz/anchor": "^0.27.0"
```
We'd like to run the tests against the latest version in repository. If you check the `yarn.lock` file, you'd see that this change makes it so that we test against the last released version on NPM instead of the latest version in the repository. Please remove this and update the lock file.

acheroncrypto replied:
Need to build ts packages first and link them to be used in tests, see https://github.com/coral-xyz/anchor/blob/master/.github/actions/setup-ts/action.yaml

We should definitely add a script for this though https://github.com/coral-xyz/anchor/issues/2471

ngundotra replied:
Thanks for this! This worked

acheroncrypto commented on tests/events/tests/events.js:
```diff
@@ -54,6 +55,34 @@ describe("events", () => {
   });
+
+  it("Self-CPI events work", async () => {
+    await sleep(200);
```
Do we need to sleep here?

ngundotra replied:
Yes

acheroncrypto reviewed (commented):
Thank you for addressing previous comments. I've left a few more comments, some of them are pretty important. Let me know what you think!

ngundotra committed:
optimize emit macro to reduce cloning

ngundotra commented on lang/attribute/event/src/lib.rs:
```diff
@@ -81,6 +81,35 @@ pub fn emit(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+            vec![anchor_lang::solana_program::instruction::AccountMeta::new_readonly(*program_info.key, false)],
+        );
+        anchor_lang::solana_program::program::invoke(&ix, &[program_info.clone()])
+            .map_err(|e: anchor_lang::solana_program::program_error::ProgramError| crate::Error::from(e))?;
```
Fixed

ngundotra committed:
explicitly only parse two args

ngundotra committed:
update events package.json

ngundotra committed:
add event instruction error code to anchor

ngundotra committed:
add event authority

ngundotra committed:
require event authority PDA to sign

ngundotra committed:
turn on seeds to hide eventAuthority

ngundotra committed:
change feature to cpi-events

ngundotra committed:
fix no-idl, no-cpi-events, and cpi-events features

ngundotra committed:
update tests

ngundotra committed:
fix no-idl cfg dispatch

ngundotra committed:
fix tests/events

ngundotra committed:
remove cpi-events from Anchor.toml

ngundotra committed:
add documentation

ngundotra committed:
slightly better interface for self-program in ctx

ngundotra commented:
FWIW I've attempted to write up an explanation of CPI Events here, along with additional tests: https://github.com/ngundotra/srfc-event-indexing

acheroncrypto commented on lang/syn/src/codegen/program/dispatch.rs:
```diff
@@ -67,17 +67,38 @@ pub fn generate(program: &Program) -> proc_macro2::TokenStream {
+                   // If the method identifier is the event tag, then execute an event cpi
+                   // against the noop instruction injected into all Anchor programs unless they have
+                   // no-cpi-events enabled.
+                    #[cfg(not(feature = "no-cpi-events"))]
```
What would you say to making this feature opt-in instead of opt-out? I think it makes sense that IDL feature is opt-out since pretty much everyone uses Anchor IDLs but I don't think we can say the same for this feature.

ngundotra replied:
Yeah, I think it makes more sense to have this as `opt-in`. Do you want to make these changes?

acheroncrypto replied:
Sure, will do after reducing the repetition.

acheroncrypto replied:
Done in https://github.com/coral-xyz/anchor/pull/2438/commits/eb051efdb3ea24a376a4faa62e2f19efbc812ecc.

acheroncrypto commented on tests/events/Anchor.toml:
```diff
@@ -1,3 +1,6 @@
+[features]
+seeds = true
```
I think instead of using opt-in feature `seeds`, we can supply these accounts when needed from client side.

acheroncrypto commented on tests/events/programs/events/src/lib.rs:
```diff
@@ -31,6 +44,14 @@ pub struct Initialize {}
+pub struct TestEventCpi<'info> {
+    program: Program<'info, crate::program::Events>,
+    /// CHECK: this is the global event authority
+    #[account(seeds=[b"__event_authority"], bump)]
```
Specifying these accounts for all instructions in a program is going to be really cumbersome especially with `__event_authority` as seed. `__` is used for internal usage but we are exposing it to outside world here. And also the seed cannot be changed because that's being used inside the macro to sign the transaction anyway.

ngundotra replied:
Yeah I agree this is a huge PITA. I'm looking forward to seeing some new ideas on this :)

acheroncrypto commented on tests/events/test-validator.log:
Log file, please remove.

acheroncrypto commented on tests/events/tests/events.js:
```diff
@@ -1,4 +1,5 @@
 const anchor = require("@coral-xyz/anchor");
+const { bs58, base64 } = require("@coral-xyz/anchor/dist/cjs/utils/bytes");
```
This import could break if folder structure changes, these exports are available in `anchor.utils.bytes`.

ngundotra replied:
It should still be able to be imported like this. I tried `"@coral-xyz/anchor/utils/bytes"` and it did not resolve as a valid import path.

Maybe I'm doing something wrong ¯\_(ツ)_/¯

acheroncrypto reviewed (commented):
Hey, thanks for the improvements. The implementation works as expected but it's very cumbersome to use at this point.

- For each event call, we supply 4 arguments when the user only cares about passing the event struct.
- We are passing the same 2 accounts to each instruction(a lot of copy paste), user's shouldn't care about this.
- Client side is the same story, a lot of repetition especially if the user doesn't want to use the `seeds` feature.
- I think this feature should be opt-in instead of opt-out.

With that being said, I think this is a great starting point and I have a few ideas for how to reduce the friction for users. Please let me know if you are fine with me working on it.

ngundotra commented:
@acheroncrypto absolutely, looking forward to seeing your changes. 

Let me know if you want to delegate some work back to me on improving this.

acheroncrypto committed:
Remove accounts and bump argument

`emit_cpi` now only takes the event as an argument. Caveat: Context must be in scope and must exactly be `ctx`.

acheroncrypto committed:
Add `event_cpi` attribute macro

acheroncrypto committed:
Generate IDL accounts with `event_cpi` macro

acheroncrypto committed:
Resolve event CPI accounts in client

acheroncrypto committed:
Update tests

acheroncrypto commented:
Here is the summary for usage with my last changes:

### Instruction

```diff
pub fn my_instruction(ctx: Context<MyInstruction>) -> Result<()> {
-    emit_cpi!(
-        ctx.accounts.program.to_account_info(),
-        ctx.accounts.event_authority.clone(),
-        *ctx.bumps.get("event_authority").unwrap(),
-        MyEvent { data: 5 }
-    );
+    emit_cpi!(MyEvent { data: 5 });
    Ok(())
}
```

### Accounts

```diff
+#[event_cpi]
#[derive(Accounts)]
-pub struct MyInstruction<'info> {
-    /// CHECK: this account is needed to guarantee that your program is the one doing the logging
-    #[account(seeds=[b"__event_authority"], bump)]
-    pub event_authority: AccountInfo<'info>,
-    pub program: Program<'info, crate::program::MyProgramName>,
+pub struct MyInstruction {}
```

### Client

```diff
-await program.methods
-.myInstruction()
-.accounts({
-  eventAuthority: anchor.web3.PublicKey.findProgramAddressSync(
-    [Buffer.from("__event_authority")],
-    program.programId
-  )[0],
-  program: program.programId,
-})
-.rpc();
+await program.methods.myInstruction().rpc();
```

acheroncrypto commented on lang/syn/src/codegen/program/handlers.rs:
```diff
@@ -91,6 +91,16 @@ pub fn generate(program: &Program) -> proc_macro2::TokenStream {
+        quote! {
+            #[inline(never)]
+            #[cfg(not(feature = "no-cpi-events"))]
+            pub fn __event_dispatch(program_id: &Pubkey, accounts: &[AccountInfo], event_data: &[u8]) -> anchor_lang::Result<()> {
```
We are not validating anything here, anyone can call this instruction.

acheroncrypto replied:
Added validation in https://github.com/coral-xyz/anchor/pull/2438/commits/87652d2f90ae00788482fbd540aff3d8be56a386.

acheroncrypto commented on lang/syn/src/parser/accounts/mod.rs:
```diff
@@ -20,23 +20,91 @@ pub fn parse(strct: &syn::ItemStruct) -> ParseResult<AccountsStruct> {
+            #fields
+
+            /// CHECK: Only the event authority can call the CPI event
+            #[account(seeds = [b"__event_authority"], bump)]
```
Validation of this account should be in the event CPI handler instead of here.

acheroncrypto commented on lang/syn/src/parser/accounts/mod.rs:
```diff
@@ -20,23 +20,91 @@ pub fn parse(strct: &syn::ItemStruct) -> ParseResult<AccountsStruct> {
+            pub event_authority: AccountInfo<#info_lifetime>,
+            /// CHECK: The program itself
+            #[account(address = crate::ID)]
+            pub program: AccountInfo<#info_lifetime>,
```
Do we even need to pass the program?

acheroncrypto replied:
Removed the `#[account(address = crate::ID)]` check because we are using `crate::ID` as the `program_id` on self-CPI.

acheroncrypto committed:
Fix clippy

acheroncrypto committed:
Remove accounts from test

acheroncrypto committed:
Remove Anchor.toml features in tests

acheroncrypto committed:
Add malicious invocation test

acheroncrypto committed:
Validate authority in the self-cpi handler to block malicious invocations

acheroncrypto committed:
Make `event-cpi` feature opt-in instead of opt-out

acheroncrypto commented:
Last changes:
- Anyone was able to invoke the event handler before 87652d2. We're now validating the event authority signed the transaction and is correct key in the event handler instruction, added a test case for this so that only the program itself can invoke the event handler.
- Removed the `#[account(address = crate::ID)]` check from the main instruction because we're invoking with `crate::ID` as the program id and it will fail if the program wasn't passed in.
- Renamed the `no-cpi-events`feature to `event-cpi` and made the feature opt-in instead of opt-out.

@ngundotra let me know what you think.

acheroncrypto committed:
Fix parsing multiple fields

ngundotra commented:
This works great! The `emit_cpi!{}` macro just works, and the `#[event_cpi]` attr is great.

I think we need 2 changes to the JS sdk tho.

1. We need a method like `program.cpiEvents.parseTransaction()` and gives a list of events from a transaction
2. Probably the `eventAuthority` address derivation on `program.addresses.eventAuthority` (and probably export IDL similarly like `program.addresses.idl`) - probably can be done in another PR

What do you think?

acheroncrypto committed:
Generate attributes and fields inside the main `TokenStream`

acheroncrypto committed:
Add documentation

acheroncrypto committed:
Add a note about `ctx` being in scope

acheroncrypto commented:
> I think we need 2 changes to the JS sdk tho.
> 
> 1. We need a method like `program.cpiEvents.parseTransaction()` and gives a list of events from a transaction
> 2. Probably the `eventAuthority` address derivation on `program.addresses.eventAuthority` (and probably export IDL similarly like `program.addresses.idl`) - probably can be done in another PR
> 
> What do you think?

Yeah client side needs improvement and I think both of those can be made in a separate PR.

One potential issue here is that we are deriving the key each time the event handler is called to validate the authority(on top of deriving it on the main instruction) but the authority key and bump does not change and can be hard coded inside the program. The solution would most likely be a similar macro to `declare_id!` which kind of sucks. I'll come back to this if it causes noticable difference in compute units.

I've added and updated the documentation and looking to merge this soon if it looks good to you as well.

ngundotra commented:
Awesome, that sounds like a plan.

Thanks for updating documentation, your changes look great to me. I tested externally by importing in another standalone repo, and it works like a charm.

🚢  it

acheroncrypto committed:
Merge master

acheroncrypto committed:
Update CHANGELOG
